torch_brain/models/poyo.py

```diff
@@ -0,0 +1,354 @@
+                "`model.unit_emb.initialize_vocab(unit_ids)`"
+            )
+
+        if self.session_emb.is_lazy():
+            raise ValueError(
+                "Session vocabulary has not been initialized, please use "
+                "`model.session_emb.initialize_vocab(session_ids)`"
+            )
+
+        # input
+        inputs = self.unit_emb(input_unit_index) + self.token_type_emb(input_token_type)
+        input_timestamp_emb = self.rotary_emb(input_timestamps)
+
+        # latents
+        latents = self.latent_emb(latent_index)
+        latent_timestamp_emb = self.rotary_emb(latent_timestamps)
+
+        # outputs
+        output_queries = self.session_emb(output_session_index)
+        output_timestamp_emb = self.rotary_emb(output_timestamps)
+
+        # encode
+        latents = latents + self.enc_atn(
+            latents,
+            inputs,
+            latent_timestamp_emb,
+            input_timestamp_emb,
+            input_mask,
+        )
+        latents = latents + self.enc_ffn(latents)
+
+        # process
+        for self_attn, self_ff in self.proc_layers:
+            latents = latents + self.dropout(self_attn(latents, latent_timestamp_emb))
+            latents = latents + self.dropout(self_ff(latents))
+
+        # decode
+        output_queries = output_queries + self.dec_atn(
+            output_queries,
+            latents,
+            output_timestamp_emb,
+            latent_timestamp_emb,
+        )
+        output_latents = output_queries + self.dec_ffn(output_queries)
+        output = self.readout(output_latents)
+
+        return output
+
+
+def poyo_mp(dim_out, ckpt_path=None):
+    if ckpt_path is not None:
+        raise NotImplementedError("Loading from checkpoint is not supported yet.")
+
+    return POYO(
+        dim=64,
+        dim_out=dim_out,
+        dim_head=64,
+        num_latents=16,
+        depth=6,
+        cross_heads=2,
+        self_heads=8,
+        ffn_dropout=0.2,
+        lin_dropout=0.4,
+        atn_dropout=0.2,
+    )
+
+
+class POYOTokenizer:
+    r"""Tokenizer used to tokenize Data for the POYO model.
+
+    This tokenizer can be called as a transform. If you are applying multiple
+    transforms, make sure to apply this one last.
+
+    Args:
+        unit_tokenizer (Callable): Tokenizer for the units.
+        session_tokenizer (Callable): Tokenizer for the sessions.
+        decoder_registry (Dict): Registry of the decoders.
+        weight_registry (Dict): Registry of the weights.
+        latent_step (float): Step size for generating latent tokens.
+        num_latents_per_step (int): Number of latents per step.
+        subtask_weights (List[float]): Loss-weights for different subtasks.
+    """
+
+    def __init__(
+        self,
+        unit_tokenizer: Callable,
+        session_tokenizer: Callable,
+        latent_step: float,
+        num_latents_per_step: int,
+        modality_spec: ModalitySpec,
```

Comment:
should we name it `readout_spec`? `modality_spec` is more consistent, but could be confusing since spikes itself is a modality. and in the future we can change the input modality. so for consistency I would suggest calling this `readout_spec`

Reply:
We should change this in the spec python file as well in that case. I just went with whatever name was used there. I support renaming it to readout_spec at both places.

Reply:
what do you mean by the spec python file? you mean renaming `ModalitySpec` to `ReadoutSpec`?

Reply:
Yes

Reply:
Renamed `modality_spec` to `readout_spec`. Done.